asterius/src/Asterius/Types.hs

```diff
@@ -140,6 +145,30 @@ instance Semigroup AsteriusCachedModule where
 instance Monoid AsteriusCachedModule where
   mempty = AsteriusCachedModule mempty mempty
 
+instance GHC.Binary AsteriusCachedModule where
+  get bh = do
+    getObjectMagic bh
+    dependencyMap <- GHC.get bh
+    fromCachedModule <- GHC.get bh
+    pure AsteriusCachedModule {..}
+
+  put_ bh AsteriusCachedModule {..} = do
+    putObjectMagic bh
+    GHC.put_ bh dependencyMap
+    GHC.put_ bh fromCachedModule
+
+objectMagic :: BS.ByteString
+objectMagic = "!<asterius>\n"
+
+putObjectMagic :: GHC.BinHandle -> IO ()
+putObjectMagic bh = for_ (BS.unpack objectMagic) (GHC.putByte bh)
```

Comment:
Using `putByte`/`getByte` is inefficient, we should use the `Binary` instance of `ByteString` and call `put_`/`get` directly.

Reply:
Hmmm, the reason I avoided it is because the instance for `ByteString` uses a length-prefixed encoding, which makes it non constant-time. If it is a malformed object file, we could read any kind of length there. I'd use `peek`/`poke` (e.g. as for the [`Word` instances](https://haskell-code-explorer.mfix.io/package/ghc-8.6.5/show/utils/Binary.hs#L1275)) but this part of the interface is not exposed.

Reply:
In that case, we should still use `putPrim`/`getPrim`. For putting, use `Data.ByteString.Unsafe.unsafeUseAsCStringLen` to expose the internal pointer/length of `objectMagic` and then use `putPrim` to put it; for `getPrim`, after getting the pointer, no need to reconstruct a `ByteString` for comparing against `objectMagic`, since we can `memcmp` directly.

Reply:
Neither `putPrim` nor `getPrim` is exported from `Binary` though :/

Reply:
Just noticed that. Yeah, then the current implementation is fine as it is.